### ISSUE TITLE
Non-returning default singledispatch function flagged as an error

### ISSUE THREAD

**Issue by nicktimko** [PyCQA/pylint#2155]
### Steps to reproduce

When using `functools.singledispatch` to delegate by type and where the default behavior (the function decorated with `@singledispatch`) for unsupported types is to raise an error, Pylint thinks it's guaranteed to fail.

```python
# singledis.py
from functools import singledispatch

def myfunc(x):
    y = f(x)
    return y

@singledispatch
def f(x):
    raise TypeError('invalid')

@f.register(str)
def f_str(x):
    return x

print(myfunc('hello'))
```

### Current behavior

```bash
$ pylint -E singledis.py              
No config file found, using default configuration                                    
************* Module singledis                                                       
E:  5, 4: Assigning to function call which doesn't return (assignment-from-no-return)
                                  
$ python singledis.py  # function call returned OK
hello                                                                                
```

### Expected behavior

No error raised.

### pylint --version output

* pylint 1.9.1,                                
* astroid 1.6.4                                
* Python 3.6.5 (default, Apr 27 2018, 11:25:39)
* [GCC 5.4.0 20160609]

**Comment by PCManticore** [comment]
This makes sense, thank you!

**Comment by Pierre-Sassoulas** [comment]
Closing as this does not raise any message as of 2.14.
